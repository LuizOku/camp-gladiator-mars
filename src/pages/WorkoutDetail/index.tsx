import React, { useEffect, useState } from "react";

import ReactPlayer from "react-player";
import { useParams } from "react-router-dom";

import { getWorkouts } from "../../api/workout";
import { useAppSelector } from "../../hooks/useAppSelector";
import { useAppDispatch } from "../../hooks/useAppDispatch";
import { useDeviceWidth } from "../../hooks/useDeviceWidth";
import {
  setList as setWorkoutList,
  selectWorkoutList,
} from "../../slices/workoutSlice";
import {
  Container,
  Title,
  Description,
  VideoWrapper,
  TagsContainer,
  StyledTag,
} from "./styles.css";

const WorkoutDetail = () => {
  const deviceWidth = useDeviceWidth();
  const { workoutId } = useParams();
  const dispatch = useAppDispatch();
  const workouts = useAppSelector(selectWorkoutList);
  const workout = workouts.find((wor) => wor.id === workoutId);
  const [screenSizeMultiplier, setScreenSizeMultiplier] = useState(2);

  useEffect(() => {
    if (deviceWidth < 650) {
      setScreenSizeMultiplier(1);
    } else {
      setScreenSizeMultiplier(2);
    }
  }, [deviceWidth]);

  useEffect(() => {
    const fetchWorkouts = async () => {
      if (workouts.length > 0) {
        return;
      }
      const res = await getWorkouts();
      dispatch(setWorkoutList(res));
    };
    fetchWorkouts();
  }, [workouts]);

  return (
    <Container>
      <Title type="h3" variation="display" font="gotham">
        {workout?.title}
      </Title>
      <Description>{workout?.description}</Description>
      <TagsContainer isMobile={screenSizeMultiplier === 1}>
        <StyledTag
          size={screenSizeMultiplier === 2 ? "default" : "small"}
          emphasis="primary"
          variation="alternative"
        >
          Level {workout?.levelTag}
        </StyledTag>
        <StyledTag
          size={screenSizeMultiplier === 2 ? "default" : "small"}
          emphasis="default"
        >
          Impact {workout?.impactTag}
        </StyledTag>
        <StyledTag
          size={screenSizeMultiplier === 2 ? "default" : "small"}
          emphasis="default"
          variation="alternative"
        >
          {workout?.duration} mins
        </StyledTag>
      </TagsContainer>
      <VideoWrapper>
        <ReactPlayer
          data-testid="video-player"
          width={320 * screenSizeMultiplier}
          height={180 * screenSizeMultiplier}
          url={workout?.media}
          playing={true}
          controls={true}
        />
      </VideoWrapper>
    </Container>
  );
};

export default WorkoutDetail;
