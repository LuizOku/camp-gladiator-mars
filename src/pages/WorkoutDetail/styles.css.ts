import styled from "styled-components";

export const Container = styled.div`
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  background-color: #f9f9f9;
  padding: 100px;
`;
