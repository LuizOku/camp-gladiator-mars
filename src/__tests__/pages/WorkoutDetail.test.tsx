import { render, screen, act } from "@testing-library/react";
import { MemoryRouter as Router } from "react-router-dom";
import { Provider } from "react-redux";

import WorkoutDetail from "../../pages/WorkoutDetail";
import { setupStore } from "../../store";
import "../../setupTests";

const workouts = [
  {
    description:
      "Enjoy a fun and high energy workout with Amber. We will engage a full body workout and get your heart rate up!",
    thumbnail:
      "https://cgcdn.s3.amazonaws.com/nation/users/UID-W1405610919UID-53c7eba7a992b.png",
    levelTag: "beginner",
    media: "https://virtual-library.s3.amazonaws.com/Week2-AmberAutrey.mp4",
    title: "Strength and Agility by Amber Autrey",
    impactTag: "high",
    id: "3",
    trainerId: "3",
    duration: 60,
  },
  {
    description:
      "Enjoy a fun and low impact workout with Claire. We will engage a full body workout and get your heart rate up!",
    thumbnail:
      "https://cgcdn.s3.amazonaws.com/nation/users/UID-D1489527062UID-58c86116311d9.png",
    levelTag: "beginner",
    media: "https://virtual-library.s3.amazonaws.com/Week2-ClaireDonahoe.mp4",
    title: "Strength and Agility by Claire Donahoe",
    impactTag: "low",
    id: "4",
    trainerId: "4",
    duration: 60,
  },
];

const promise = Promise.resolve();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({
    workoutId: "3",
  }),
}));

const renderComponent = () => {
  const store = setupStore({
    workout: {
      list: workouts,
    },
  });
  return render(
    <Provider store={store}>
      <Router>
        <WorkoutDetail />
      </Router>
    </Provider>
  );
};

describe("<WorkoutDetail />", () => {
  it("renders without crashing", async () => {
    const view = renderComponent();
    expect(view).toMatchSnapshot();
    await act(async () => {
      await promise;
    });
  });

  it("should have the workout title", async () => {
    renderComponent();
    expect(screen.getByText(workouts[0].title)).toBeInTheDocument();
    await act(async () => {
      await promise;
    });
  });

  it("should have the workout description", async () => {
    const promise = Promise.resolve();
    renderComponent();
    expect(screen.getByText(workouts[0].description)).toBeInTheDocument();
    await act(async () => {
      await promise;
    });
  });

  it("should have the workout level tag", async () => {
    renderComponent();
    expect(
      screen.getByText(`Level ${workouts[0].levelTag}`)
    ).toBeInTheDocument();
    await act(async () => {
      await promise;
    });
  });

  it("should have the workout impact tag", async () => {
    renderComponent();
    expect(
      screen.getByText(`Impact ${workouts[0].impactTag}`)
    ).toBeInTheDocument();
    await act(async () => {
      await promise;
    });
  });

  it("should have the workout duration tag", async () => {
    renderComponent();
    expect(
      screen.getByText(`${workouts[0].duration} mins`)
    ).toBeInTheDocument();
    await act(async () => {
      await promise;
    });
  });

  it("should have the workout video", async () => {
    renderComponent();
    expect(screen.getByTestId("video-player")).toBeInTheDocument();
    await act(async () => {
      await promise;
    });
  });
});
