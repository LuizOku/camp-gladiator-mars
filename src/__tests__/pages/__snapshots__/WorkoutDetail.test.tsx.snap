// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WorkoutDetail /> renders without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bczRLJ brMoBP"
      >
        <h3
          class="heading sc-gsnTZi lmBgFI"
          data-font="gotham"
          data-variation="display"
        >
          Strength and Agility by Amber Autrey
        </h3>
        <span
          class="label sc-dkzDqf bmHYPV"
          data-size="default"
        >
          Enjoy a fun and high energy workout with Amber. We will engage a full body workout and get your heart rate up!
        </span>
        <div
          class="sc-eCYdqJ gUGBKD"
        >
          <div
            class="tag sc-jSMfEi dxkJpK"
            data-emphasis="primary"
            data-size="default"
            data-variation="alternative"
          >
            <p
              class="paragraph "
              data-size="small"
              data-weight="bold"
            >
              Level 
              beginner
            </p>
          </div>
          <div
            class="tag sc-jSMfEi dxkJpK"
            data-emphasis="default"
            data-size="default"
            data-variation="default"
          >
            <p
              class="paragraph "
              data-size="small"
              data-weight="bold"
            >
              Impact 
              high
            </p>
          </div>
          <div
            class="tag sc-jSMfEi dxkJpK"
            data-emphasis="default"
            data-size="default"
            data-variation="alternative"
          >
            <p
              class="paragraph "
              data-size="small"
              data-weight="bold"
            >
              60
               mins
            </p>
          </div>
        </div>
        <div
          class="sc-hKMtZM cbTIWG"
        >
          <div
            data-testid="video-player"
            style="width: 640px; height: 360px;"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bczRLJ brMoBP"
    >
      <h3
        class="heading sc-gsnTZi lmBgFI"
        data-font="gotham"
        data-variation="display"
      >
        Strength and Agility by Amber Autrey
      </h3>
      <span
        class="label sc-dkzDqf bmHYPV"
        data-size="default"
      >
        Enjoy a fun and high energy workout with Amber. We will engage a full body workout and get your heart rate up!
      </span>
      <div
        class="sc-eCYdqJ gUGBKD"
      >
        <div
          class="tag sc-jSMfEi dxkJpK"
          data-emphasis="primary"
          data-size="default"
          data-variation="alternative"
        >
          <p
            class="paragraph "
            data-size="small"
            data-weight="bold"
          >
            Level 
            beginner
          </p>
        </div>
        <div
          class="tag sc-jSMfEi dxkJpK"
          data-emphasis="default"
          data-size="default"
          data-variation="default"
        >
          <p
            class="paragraph "
            data-size="small"
            data-weight="bold"
          >
            Impact 
            high
          </p>
        </div>
        <div
          class="tag sc-jSMfEi dxkJpK"
          data-emphasis="default"
          data-size="default"
          data-variation="alternative"
        >
          <p
            class="paragraph "
            data-size="small"
            data-weight="bold"
          >
            60
             mins
          </p>
        </div>
      </div>
      <div
        class="sc-hKMtZM cbTIWG"
      >
        <div
          data-testid="video-player"
          style="width: 640px; height: 360px;"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
